backend/kale/common/cephfsutils.py

```diff
@@ -0,0 +1,258 @@
+
+
+def snapshot_pod():
+    """Take snapshots of the current Pod's PVCs."""
+    volumes = [(path, volume.name, size)
+            for path, volume, size in podutils.list_volumes()]
+    namespace = podutils.get_namespace()
+    pod_name = podutils.get_pod_name()
+    log.info("Taking a snapshot of pod %s in namespace %s ..."
+             % (pod_name, namespace))
+    snapshot_names = []
+    for i in volumes:
+        snapshot_pvc("snapshot." + i[1], i[1], pod=pod_name, 
+        default_container=podutils.get_container_name())
+        snapshot_names.append("snapshot." + i[1])
+    return snapshot_names
+
+
+def snapshot_notebook():
+    """Take snapshots of the current Notebook's PVCs and store its metadata."""
+    volumes = [(path, volume.name, size)
+            for path, volume, size in podutils.list_volumes()]
+    namespace = podutils.get_namespace()
+    pod_name = podutils.get_pod_name()
+    log.info("Taking a snapshot of notebook %s in namespace %s ..."
+             % (pod_name, namespace))
+    snapshot_names = []
+    for i in volumes:
+        snapshot_pvc("snapshot." + i[1], i[1],
+        image=podutils.get_docker_base_image(),
+        path=i[0],
+        pod=pod_name,
+        default_container=podutils.get_container_name(),
+        is_workspace_dir=str(podutils.is_workspace_dir(i[0])))
+        snapshot_names.append("snapshot." + i[1])
+    return snapshot_names
+
+
+def check_snapshot_status(snapshot_name):
+    """Check if volume snapshot is ready to use."""
+    log.info("Checking snapshot with snapshot name: %s", snapshot_name)
+    task = None
+    status = None
+    task = get_pvc_snapshot(snapshot_name=snapshot_name)
+    status = task['status']['readyToUse']
+
+    if status == True:
+        log.info("Successfully created volume snapshot")
+    elif status == False:
+        raise RuntimeError("Snapshot not ready (status: %s)" % status)
+    else:
+        raise RuntimeError("Unknown snapshot task status: %s" % status)
+    return status
+
+
+def get_pvc_snapshot(snapshot_name):
+    """Get info about a pvc snapshot."""
+    co_client = podutils._get_k8s_custom_objects_client()
+    namespace = podutils.get_namespace()
+    
+    pvc_snapshot = co_client.get_namespaced_custom_object(
+    group="snapshot.storage.k8s.io",
+    version="v1beta1",
+    namespace=namespace,
+    plural="volumesnapshots",
+    name=snapshot_name,
+    )
+    return pvc_snapshot
+
+
+def list_pvc_snapshots(label_selector=""):
+    """List pvc snapshots."""
+    co_client = podutils._get_k8s_custom_objects_client()
+    namespace = podutils.get_namespace()
+    
+    pvc_snapshots = co_client.list_namespaced_custom_object(
+    group="snapshot.storage.k8s.io",
+    version="v1beta1",
+    namespace=namespace,
+    plural="volumesnapshots",
+    label_selector=label_selector)
+    return pvc_snapshots
+
+
+def hydrate_pvc_from_snapshot(new_pvc_name, source_snapshot_name):
+    """Create a new PVC out of a Rok snapshot."""
```

Comment:
Change Rok to Ceph

Reply:
Thanks for catching that. I changed `Rok` to `volume`, to keep it more general. What does need some closer attention in this part is the `access_mode`, which I chose to used `ReadWriteMany` while the original Rok code uses `ReadWriteOnce`. Rok needs ReadWriteMany for certain things so I chose to use that here, though I am not sure if it is required. This also limits the supported CSI storage providers to those that support ReadWriteMany. I think the best solution is to create a function and check if the CSI supports ReadWriteMany (this could then also be used when generating the pipeline so people do not need to edit the python as stated [here](https://github.com/kubeflow-kale/kale/blob/master/FAQ.md#pod-has-unbound-immediate-persistentvolumeclaim).